service/history/replication/sync_state_retriever.go

```diff
@@ -0,0 +1,429 @@
+	if err != nil {
+		s.logger.Error("SyncWorkflowState failed to get current version history and update progress cache",
+			tag.WorkflowNamespaceID(namespaceID),
+			tag.WorkflowID(execution.WorkflowId),
+			tag.WorkflowRunID(execution.RunId),
+			tag.Error(err))
+		return result, nil
+	}
+	result.LastVersionHistory = currentVersionHistory
+	return result, nil
+}
+
+func (s *SyncStateRetrieverImpl) getNewRunInfo(ctx context.Context, namespaceId namespace.ID, execution *commonpb.WorkflowExecution, newRunId string) (_ *replicationpb.NewRunInfo, retError error) {
+	wfCtx, releaseFunc, err := s.workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		s.shardContext,
+		namespaceId,
+		&commonpb.WorkflowExecution{
+			WorkflowId: execution.WorkflowId,
+			RunId:      newRunId,
+		},
+		locks.PriorityLow,
+	)
+	defer func() {
+		if releaseFunc != nil {
+			releaseFunc(retError)
+		}
+	}()
+
+	if err != nil {
+		return nil, err
+	}
+	mu, err := wfCtx.LoadMutableState(ctx, s.shardContext)
+	switch err.(type) {
+	case nil:
+	case *serviceerror.NotFound:
+		s.logger.Info(fmt.Sprintf("SyncWorkflowState new run not found, newRunId: %v", newRunId),
+			tag.WorkflowNamespaceID(namespaceId.String()),
+			tag.WorkflowID(execution.WorkflowId),
+			tag.WorkflowRunID(execution.RunId))
+		return nil, nil
+	default:
+		return nil, err
+	}
+	versionHistory, err := versionhistory.GetCurrentVersionHistory(mu.GetExecutionInfo().VersionHistories)
+	if err != nil {
+		return nil, err
+	}
+	versionHistory = versionhistory.CopyVersionHistory(versionHistory)
+	releaseFunc(nil)
+	releaseFunc = nil
+	newRunEvents, err := s.getEventsBlob(ctx, versionHistory.BranchToken, common.FirstEventID, common.FirstEventID+1)
+	switch err.(type) {
+	case nil:
+	case *serviceerror.NotFound:
+		s.logger.Info(fmt.Sprintf("SyncWorkflowState new run event not found, newRunId: %v", newRunId),
+			tag.WorkflowNamespaceID(namespaceId.String()),
+			tag.WorkflowID(execution.WorkflowId),
+			tag.WorkflowRunID(execution.RunId))
+		return nil, nil
+	default:
+		return nil, err
+	}
+	if len(newRunEvents) == 0 {
+		s.logger.Info(fmt.Sprintf("SyncWorkflowState new run event is empty, newRunId: %v", newRunId),
+			tag.WorkflowNamespaceID(namespaceId.String()),
+			tag.WorkflowID(execution.WorkflowId),
+			tag.WorkflowRunID(execution.RunId))
+		return nil, nil
+	}
+	return &replicationpb.NewRunInfo{
+		RunId:      newRunId,
+		EventBatch: newRunEvents[0],
+	}, nil
+}
+
+func (s *SyncStateRetrieverImpl) getMutation(mutableState workflow.MutableState, versionedTransition *persistencepb.VersionedTransition) (*persistencepb.WorkflowMutableStateMutation, error) {
+	rootNode := mutableState.HSM()
+	updatedStateMachine, err := s.getUpdatedSubStateMachine(rootNode, versionedTransition)
+	if err != nil {
+		return nil, err
+	}
+	mutableStateClone := mutableState.CloneToProto()
+	err = workflow.SanitizeMutableState(mutableStateClone)
```

Comment:
Since this logic is within mutable state lock, I think ideally we only want to clone, sanitize, & drop unknown fields for the returned mutation. We probably need another SanitizeMutableStateMutation method.